controllers/apps/components/base_rsm.go

```diff
@@ -0,0 +1,742 @@
+			removeFinalizerPVC.SetFinalizers([]string{})
+			removeFinalizerPVCVertex := c.PatchResource(removeFinalizerPVC, pvc, fromVertex)
+			return c.DeleteResource(pvc, removeFinalizerPVCVertex)
+		},
+		removePVClaimRefStep: func(fromVertex *ictrltypes.LifecycleVertex, step pvcRecreateStep) *ictrltypes.LifecycleVertex {
+			// step 3: remove claimRef in pv
+			removeClaimRefPV := pv.DeepCopy()
+			if removeClaimRefPV.Spec.ClaimRef != nil {
+				removeClaimRefPV.Spec.ClaimRef.UID = ""
+				removeClaimRefPV.Spec.ClaimRef.ResourceVersion = ""
+			}
+			return c.PatchResource(removeClaimRefPV, pv, fromVertex)
+		},
+		createPVCStep: func(fromVertex *ictrltypes.LifecycleVertex, step pvcRecreateStep) *ictrltypes.LifecycleVertex {
+			// step 4: create new pvc
+			newPVC.SetResourceVersion("")
+			return c.CreateResource(newPVC, fromVertex)
+		},
+		pvRestorePolicyStep: func(fromVertex *ictrltypes.LifecycleVertex, step pvcRecreateStep) *ictrltypes.LifecycleVertex {
+			// step 5: restore to previous pv policy
+			restorePV := pv.DeepCopy()
+			policy := corev1.PersistentVolumeReclaimPolicy(restorePV.Annotations[constant.PVLastClaimPolicyAnnotationKey])
+			if len(policy) == 0 {
+				policy = corev1.PersistentVolumeReclaimDelete
+			}
+			restorePV.Spec.PersistentVolumeReclaimPolicy = policy
+			return c.PatchResource(restorePV, pv, fromVertex)
+		},
+	}
+
+	updatePVCByRecreateFromStep := func(fromStep pvcRecreateStep) {
+		lastVertex := c.WorkloadVertex
+		for step := pvRestorePolicyStep; step >= fromStep && step >= pvPolicyRetainStep; step-- {
+			lastVertex = addStepMap[step](lastVertex, step)
+		}
+	}
+
+	targetQuantity := vctProto.Spec.Resources.Requests[corev1.ResourceStorage]
+	if pvcNotFound && !pvNotFound {
+		// this could happen if create pvc step failed when recreating pvc
+		updatePVCByRecreateFromStep(removePVClaimRefStep)
+		return nil
+	}
+	if pvcNotFound && pvNotFound {
+		// if both pvc and pv not found, do nothing
+		return nil
+	}
+	if reflect.DeepEqual(pvc.Spec.Resources, newPVC.Spec.Resources) && pv.Spec.PersistentVolumeReclaimPolicy == corev1.PersistentVolumeReclaimRetain {
+		// this could happen if create pvc succeeded but last step failed
+		updatePVCByRecreateFromStep(pvRestorePolicyStep)
+		return nil
+	}
+	if pvcQuantity := pvc.Spec.Resources.Requests[corev1.ResourceStorage]; !viper.GetBool(constant.CfgRecoverVolumeExpansionFailure) &&
+		pvcQuantity.Cmp(targetQuantity) == 1 && // check if it's compressing volume
+		targetQuantity.Cmp(*pvc.Status.Capacity.Storage()) >= 0 { // check if target size is greater than or equal to actual size
+		// this branch means we can update pvc size by recreate it
+		updatePVCByRecreateFromStep(pvPolicyRetainStep)
+		return nil
+	}
+	if pvcQuantity := pvc.Spec.Resources.Requests[corev1.ResourceStorage]; pvcQuantity.Cmp(vctProto.Spec.Resources.Requests[corev1.ResourceStorage]) != 0 {
+		// use pvc's update without anything extra
+		c.UpdateResource(newPVC, c.WorkloadVertex)
```

Comment:
Will this branch trigger a volume expansion ? Should we check if the CSI support the 'allowVolumeExpansion'.

Reply:
out of scope of this pr. but I think the answer is no as volume expansion has been handled in previous block(L397-L403).